338 - SigmaBoolean serialization

ApexTheory opened the pull request:
Hey @greenhat please take a look at your convenience.  I used the existing rust & scala code as a reference to fill in the gaps.  

I do have some questions regarding the Arbitrary impl and the roundtrip test, I will add them alongside the relevant code in the PR.

EDIT(@greenhat): Close #338

ApexTheory commented on ergotree-ir/src/sigma_protocol/sigma_boolean.rs:
```diff
@@ -196,6 +211,22 @@ mod arbitrary {
+        }
+    }
+
     impl Arbitrary for SigmaBoolean {
```
There was an existing SigmaBoolean Arbitrary impl here that uses the ProveDlog enum variant.  
Do we implementations for the other variants (ProveDHTuple, Cand, Cor, Cthreshold)

greenhat replied:
Good question! Please, extend `SigmaBoolean` impl with DH tuple `Cand` and `Cor` conjectures and adding `Arbitrary` impl for them along the way. 
Be aware that `SigmaBoolean` is extensively used in PBT tests (through `SigmaProp`) for boxes, transaction building and signing, etc. Watch for the impact on overall test suite run time. We need to keep it reasonable. :)

ApexTheory replied:
Would something like this work?   
`    
        fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
            prop_oneof![
                (any::<ProveDlog>()),
                (any::<ProveDhTuple>()),
                (any::<Cand>()),
                (any::<Cor>()),
                (any::<Cthreshold>()),
            ]
            .prop_map(|p| match p {
                SigmaProofOfKnowledgeTree::ProveDlog(provedlog) => {
                    SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(provedlog))
                }
                SigmaProofOfKnowledgeTree::ProveDhTuple(provedhtuple) => {
                    SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDhTuple(
                        provedhtuple,
                    ))
                }
                SigmaConjecture::Cand(cand) => {
                    SigmaBoolean::SigmaConjecture(SigmaConjecture::Cand(cand))
                }
                SigmaConjecture::Cor(cor) => {
                    SigmaBoolean::SigmaConjecture(SigmaConjecture::Cor(cor))
                }
                SigmaConjecture::Cthreshold(cthreshold) => {
                    SigmaBoolean::SigmaConjecture(SigmaConjecture::Cthreshold(cthreshold))
                }
            })
            .boxed()
        }
    }
`

Also, I am still working on the `Arbitrary` impl for `Cand`/`Cor`...it seems I need to bound the `SigmaConjectureItems` vectors between 2 and 1000?  Maybe its better to keep the size small to minimize test time as you mentioned?

greenhat replied:
1. There is a `prop_recursive` for the generation of recursive structures. See how it's used in constant generator - https://github.com/ergoplatform/sigma-rust/blob/e4965fcd758c0d154bb81ec424d4c70a189a8514/ergotree-ir/src/mir/constant.rs#L313-L359 . 
2. There should be `From` implemented for `SigmaBoolean` so that would allow simple `.into()` or `.prop_map_into()`.
3. Yes, something like `2..5` should suffice for conjecture items.

ApexTheory commented on ergotree-ir/src/sigma_protocol/sigma_boolean.rs:
```diff
@@ -220,4 +251,16 @@ mod arbitrary {
+
+        #[test]
+        fn sigma_boolean_ser_roundtrip(v in any::<SigmaBoolean>()) {
+            prop_assert_eq![sigma_serialize_roundtrip(&v), v]
```
When I print out the generated SigmaBoolean value on my local machine, I ALWAYS get the following:
ProofOfKnowledge(ProveDlog(ProveDlog { h: EC:0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798 }))

Should this value not be changing as it is randomly generated?

greenhat replied:
Good catch! I've been thinking about that for a while. Although we have `DlogProverInput::random()` from which we can get a random `EcPoint` with `DlogProverInput::public_image()` it's in `ergotree-interpreter` crate. I was thinking of moving it to `ergotree-ir` but eventually declined since it belongs to an "interpreter's" part of the sigma protocol.
So I suggest to make `EcPoint::random()` (exp the group generator) and use it in `Arbitrary` impl for `EcPoint` (see `Arbitrary` impl for `DlogProverInput`).

greenhat reviewed (commented):
Looking good! Thanks for taking care of this!

ApexTheory commented:
Hi @greenhat , I pushed a lot of changes into the branch; I am now able to generate the arbitrary `SigmaBoolean` values (either a proof or conjecture based on param), but there are a few things I likely need to look at further:
1) prop_recursive params - I just defaulted everything to 1 for now and did not explore much beyond
2) should the `SigmaBoolean ` Arbitrary impl be calling the `Cand`/`Cor` Arbitary impl?
3) is use of unwrap okay within the `Cand`/`Cor` From impl?

Please take a look when you have some time, thanks!

greenhat commented on ergotree-ir/src/sigma_protocol/sigma_boolean/cand.rs:
```diff
@@ -45,11 +48,55 @@ impl Cand {
+    }
+}
+
+impl From<Vec<SigmaBoolean>> for Cand {
```
We should avoid using unwrap in production code. It's perfectly fine in tests though. 
`From` is not applicable here, since we want the error on an invalid number of items to be handled by the caller. `TryFrom` fits here, but be aware that `BoundedVec` (type behind a `SigmaConjectureItems` type alias) has one.

ApexTheory replied:
Makes sense.  So I am removing this code from `Cand`/`Cor`, and I am okay to unwrap() in the `SigmaBoolean` Arbitrary impl as below (since the caller is test code and does not need to handle the error gracefully). 
```
Cand { 
   items: BoundedVec::try_from(items).unwrap() 
}
```

Will check in both changes.

greenhat reviewed (changes requested):
Looking good!  Thank you!

> Hi @greenhat, I pushed a lot of changes into the branch (and rebased on develop); I am now able to generate the arbitrary `SigmaBoolean` values (either a proof or conjecture based on param), but there are a few things I likely need to look at further:
> 
> 1. `prop_recursive `params - I just defaulted everything to 1 for now and did not explore much beyond

I'd say this might "choke" `prop_recursive`. I suggested 16 for total element target (think of it as total nodes+leafs count of the resulted tree). `expected_branch_size` is a bit of a mystery to me, but setting it to the half of the `desired_size` seems like does the job fine. :)

> 2. ~should the SigmaBoolean Arbitrary impl be calling the Cand/Cor Arbitary impl?~ I think not

Yes, "inside" `prop_recursive`. I don't see any reason not to.

> 3. is use of unwrap okay within the `Cand`/`Cor` From impl?

No. See the reasoning in the comment.

> 4. calling `any_with::<SigmaBoolean>(ArbSigmaBooleanParams::SigmaConjecture(1))` seems to return ProofOfKnowledge values along with Conjectures, when I expect only Conjectures.  Perhaps some mechanism of `prop_oneof!` I am not understanding.

I believe it's `prop_recursive` feature to generate trees that only consist of one leaf.

> 5. impact on overall test run time

We should not significantly slow down the tests. Ideally, it should add some 5%-10%. If it gets bad, decrease `prop_recursive` parameters further.

ApexTheory commented on ergotree-ir/src/sigma_protocol/sigma_boolean.rs:
```diff
@@ -223,19 +300,72 @@ mod arbitrary {
+        fn arbitrary_with(args: Self::Parameters) -> Self::Strategy {
+            match args {
+                ArbSigmaBooleanParams::ProofOfKnowledge => primitive_type_value(),
+                ArbSigmaBooleanParams::SigmaConjecture(depth) => primitive_type_value()
```
Yes this is a good point.  I thought there may be some benefit to request only primitives via parameter.  Will update.

ApexTheory commented:
I pushed some more changes including the following:

1. updating the prop_recursive params to your suggested defaults
2. call `Cand/Cor::any` from within `SigmaBoolean`'s prop_recursive
3. removed `Cand/Cor` From impl and instead using `BoundedVec::try_from`
4. only using unwrap() in test code
5. removed `ArbSigmaBooleanParams`

One issue is that I am not able to generate nested conjectures using `any::<SigmaBoolean>()`.  It seems to be always a single conjecture with several proofs.  Need to explore the recursive parameters further.

greenhat commented on ergotree-ir/src/sigma_protocol/sigma_boolean.rs:
```diff
@@ -316,53 +315,24 @@ mod arbitrary {
+        fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
+            primitive_type_value()
+                .prop_recursive(2, 16, 8, |_elem| {
+                    prop_oneof![any::<Cand>().prop_map_into(), any::<Cor>().prop_map_into(),]
```
You don't need to call `any<Cand>()` here. `_elem` is a generator (`Strategy`) which should be used to produce a few values (via `prop::collection::vec(elem, 2..5)`) with something like this:
```rust
prop_oneof![
    prop::collection::vec(elem, 2..=4)
        .prop_map(|elems| Cand { items: elems.try_into().unwrap() }),
    prop::collection::vec(elem, 2..=4)
        .prop_map(|elems| Cor { items: elems.try_into().unwrap() })
]
```

By calling `any<Cand>()` here you're ignoring "recursive" part of `prop_recursive` producing "generic" `Cor` and `Cand`. 

This is also the reason why you cannot call `any<SigmaBoolean>()` in `Arbitrary` impl for `Cor`/`Cand`. Because in the end, it tries to call "itself".

ApexTheory replied:
Thanks for the feedback, this makes sense.  I have it working with your suggestions, will push the changes now.

greenhat commented on ergotree-ir/src/sigma_protocol/sigma_boolean/cand.rs:
```diff
@@ -59,33 +60,26 @@ impl SigmaSerializable for Cand {
-            )
-            .prop_map_into(),]
-            .boxed()
+            vec(primitive_type_value(), 2..5)
```
After the fix above you should be able to call `any<SigmaBoolean>()` here.

greenhat reviewed (changes requested):
Looking good! We're getting there, `prop_recursive` is hard! :)

greenhat reviewed (changes requested):
Looking great! I see tests are taking around 50% longer to run (comparing CI ubuntu-latest with develop). That’s a bit too much. Please try cutting down prop_recursive parameters to `2, 8, 4` and if it does not help even to `1, 8, 4`. Let’s try to get it down to 20%-30% slowdown tops.
